File: javascript/bob/bob.js
//
// This is only a SKELETON file for the "Bob" exercise. It's been provided as a
// convenience to get you started writing code faster.
//

var Bob = function () {
};

var ANSWER_DEFAULT = 'Whatever.',
    ANSWER_QUESTION = 'Sure.',
    ANSWER_NOTHING = 'Fine. Be that way!',
    ANSWER_EXCLAIM = 'Whoa, chill out!';

Bob.prototype.hey = function (input) {
    input = input.trim();
    var answer = ANSWER_DEFAULT;
    if (!input) {
        answer = ANSWER_NOTHING;
    } else if (/[A-za-z]/g.exec(input) && input.toUpperCase() === input) {
        answer = ANSWER_EXCLAIM;
    } else if (/.*[?]$/g.exec(input)) {
        answer = ANSWER_QUESTION;
    }
    return answer;
};

module.exports = Bob;

File: elixir/bob/lib/bob.ex
defmodule Bob do
  @spec hey(String.t()) :: String.t()
  def hey(input) do
    with input <- input |> String.trim(),
         is_question <- question?(input),
         is_yelling <- yelling?(input),
         is_silent <- silent?(input) do
      cond do
        is_silent -> "Fine. Be that way!"
        is_question and is_yelling -> "Calm down, I know what I'm doing!"
        is_question -> "Sure."
        is_yelling -> "Whoa, chill out!"
        true -> "Whatever."
      end
    end
  end

  defp question?(input) do
    String.ends_with?(input, "?")
  end

  defp yelling?(input) do
    with upcase <- String.upcase(input),
         gibberish <- String.downcase(input) == upcase do
      upcase == input and not gibberish
    end
  end

  defp silent?(input) do
    input == ""
  end
end
